Src/FluentAssertions/Equivalency/SelfReferenceEquivalencyAssertionOptions.cs

```diff
@@ -202,9 +202,9 @@ EqualityStrategy IEquivalencyAssertionOptions.GetEqualityStrategy(Type requested
             {
                 strategy = EqualityStrategy.ForceEquals;
             }
-            else if (type.IsRecord())
+            else if (type.IsRecord() && (compareRecordsByValue.HasValue || getDefaultEqualityStrategy is null))
```

Comment:
We can now save a dictionary lookup + some reflection on cache miss.
`(compareRecordsByValue.HasValue || getDefaultEqualityStrategy is null) && type.IsRecord()`